MSAL/src-internal/ios/MSALLegacySharedAccountsProvider.m

```diff
@@ -0,0 +1,512 @@
+                                                         withParameters:(MSALAccountEnumerationParameters *)parameters
+                                                                  error:(NSError **)error
+{
+    NSMutableArray *resultAccounts = [NSMutableArray new];
+    
+    for (NSString *accountId in [jsonDictionary allKeys])
+    {
+        NSDictionary *singleAccountDictionary = [jsonDictionary msidObjectForKey:accountId ofClass:[NSDictionary class]];
+        
+        if (!singleAccountDictionary)
+        {
+            continue;
+        }
+        
+        NSError *singleAccountError = nil;
+        MSALLegacySharedAccount *account = [MSALLegacySharedAccountFactory accountWithJSONDictionary:singleAccountDictionary error:&singleAccountError];
+        
+        if (!account)
+        {
+            MSID_LOG_WITH_CTX_PII(MSIDLogLevelWarning, nil, @"Failed to create account with error %@", MSID_PII_LOG_MASKABLE(singleAccountError));
+            continue;
+        }
+        
+        if ([account matchesParameters:parameters])
+        {
+            [resultAccounts addObject:account];
+        }
+    }
+    
+    return resultAccounts;
+}
+
+#pragma mark - Update
+
+- (BOOL)updateAccount:(id<MSALAccount>)account idTokenClaims:(NSDictionary *)idTokenClaims error:(NSError **)error
+{
+    MSID_LOG_WITH_CTX_PII(MSIDLogLevelInfo, nil, @"Updating account %@", MSID_PII_LOG_MASKABLE(account));
+    
+    return [self updateAccount:account
+                 idTokenClaims:idTokenClaims
+                tenantProfiles:nil
+                     operation:MSALLegacySharedAccountUpdateOperation
+                         error:error];
+}
+
+- (nullable NSArray<MSALLegacySharedAccount *> *)updatableAccountsFromJsonObject:(NSDictionary *)jsonDictionary
+                                                                     msalAccount:(id<MSALAccount>)msalAccount
+                                                                   idTokenClaims:(NSDictionary *)idTokenClaims
+                                                                         version:(MSALLegacySharedAccountVersion)version
+                                                                           error:(NSError **)error
+{
+    MSALAccountEnumerationParameters *parameters = [MSALLegacySharedAccountFactory parametersForAccount:msalAccount tenantProfileIdentifier:idTokenClaims[@"oid"]];
+    
+    if (!parameters)
+    {
+        NSError *parameterError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Unsupported account found, skipping update", nil, nil, nil, nil, nil);
+        [self fillAndLogError:error withError:parameterError logLine:@"Unsupported account found, skipping update"];
+        return nil;
+    }
+    
+    NSError *parseError = nil;
+    NSArray<MSALLegacySharedAccount *> *accounts = [self accountsFromJsonObject:jsonDictionary withParameters:parameters error:&parseError];
+    
+    if (parseError)
```

Comment:
nit: Should we check if (parseError && error)

Reply:
no, because error might be nil. we want to return nil regardless if developer passes error or not.